### PR TITLE
Feature/sql job stuff

### PR DESCRIPTION
**Pull request by worksofliam**
* Add new SQL Job Manager view, to maintain multiple jobs for running SQL queries with different properties
* Depend on server side component to manage multiple jobs, and provide automated installer/updated
* View to edit JDBC properties for running SQL job

#
<details>
<summary>JDBC Options</summary>

Here is a list of all the [JDBC options](https://www.ibm.com/docs/en/i/7.3?topic=ssw_ibm_i_73/rzahh/javadoc/com/ibm/as400/access/doc-files/JDBCProperties.htm)

Below is a list of all the JDBC options with any special context. Feel free to check these off as you test. There are a handful of options that take system defaults or are platform dependent. We may want to discuss with Jesse how to handle these. This PR has a lot of text, I did my best to check spelling and formatting , but am expecting to be various typos/mal formatted text etc. 

### System:
- [ ] Libraries
- [ ] Auto commit
- [ ] Concurrent access resolution
- [ ] Cursor hold
- [ ] Cursor sensitivity
- [ ] Database name
- [ ] Decfloat rounding mode
- [ ] Maximum precision
- [ ] Minimum divide scale
- [ ] Package CCSID
- [ ] Transaction isolation
- [ ] Translate hex
- [ ] True autocommit
- [ ] XA loosely coupled support
- [ ] Maximum scale

### Format
All Format properties default to "server job". For the implementation, I set the default as the first option in the list. We may need to work with Jesse on the Server component to see how we want to handle these properties.

- [ ] Naming
- [ ] Date format
- [ ] Date separator
- [ ] Decimal separator
- [ ] Time format
- [ ] Time separator

### Performance
- [ ] Big decimal
- [ ] Block criteria
- [ ] Block size
- [ ] Data compression
- [ ] Extended dynamic
- [ ] Lazy close
- [ ] LOB threshold
- [ ] Maximum blocked input rows
- [ ] Package
- [ ] Package add
- [ ] Package cache
- [ ] Package criteria
- [ ] Package error
- [ ] Package library
- [ ] Prefetch
- [ ] qaqqinilib
    - **Default**: System Default
    - default left blank in implementation 
- [ ] Query optimize goal
- [ ] Query timeout mechanism
- [ ] Receive buffer size
    - **default**: platform dependent
    - not included in implementation 
- [ ] Send buffer size
    - **default**: platform dependent
    - not included in implementation 
- [ ] Variable field compression
- [ ] Query storage limit

### Sort
- [ ] Sort
- [ ] Sort language
- [ ] Sort weight
- [ ] Sort table

### Other
- [ ] Full open
- [ ] Access
- [ ] Autocommit exception
- [ ] Bidi string type
- [ ] Bidi implicit reordering
- [ ] Bidi numeric ordering
- [ ] Data truncation
- [ ] Driver
- [ ] Errors
- [ ] Extended metadata
- [ ] Hold input locators
- [ ] Hold statements
- [ ] Ignore Warnings
- [ ] keep alive
    - **default**: platform dependent
    - not included in implementation
- [ ] Metadata source
    - **default**: "0" (V6R1 and earlier) "1" (post V6R1)
    - for this I defaulted to "1"
- [ ] Proxy Server
    - **default**: value of the `proxyServer` property, or none if not set
    - here I default to blank
- [ ] Remarks
- [ ] Secondary URL
    - **default**: current JDBC URL
    - here I default to blank
- [ ] Secure
- [ ] Server trace
- [ ] thread used
- [ ] toolbox trace
- [ ] trace
- [ ] translate binary
- [ ] translate boolean 

## Options with non static defaults
here is a dedicated list of the non static options

### Default: Server job
- [ ] Date format
- [ ] Date separator
- [ ] Decimal separator
- [ ] Time format
- [ ] Time separator

### Default: Platform dependent
- [ ] Receive buffer size
- [ ] Send buffer size
- [ ] keep alive

### Default: System default
- [ ] concurrent access resolution
- [ ] qaqqinilib

### Other non trivial defaults
- [ ] Metadata source
- [ ] Proxy Server

</details>

### CODE REVIEW

**Comment by worksofliam** [comment]
vsix inside of zip: [vscode-db2i-0.2.10.vsix.zip](https://github.com/halcyon-tech/vscode-db2i/files/11279173/vscode-db2i-0.2.10.vsix.zip)

**Comment by worksofliam** [comment]
New vsix: [vscode-db2i-0.2.10.vsix.zip](https://github.com/halcyon-tech/vscode-db2i/files/11288256/vscode-db2i-0.2.10.vsix.zip)

**Comment by worksofliam** [comment]
Next steps for this PR:

1. User testing is required
2. Merge in main for latest changes

**Comment by worksofliam** [comment]
@ajshedivy I think we should delegate @ThePrez to determine if every JDBC Option needs testing, seeing as we're using that database component, I would expect the database component to make sure they work.

**Comment by worksofliam** [comment]
Here's my user testing:

* [x] Ext. checks for database component updates when connecting to each system
    * [x] Testing disconnecting and reconnecting 
* [x] Database component is updated when a new version is available
* [x] SQL Job Manager view is shown when the DB component is installed or is already available
* [x] Am able to spin up multiple SQL jobs from the SQL Job Manager view
* [x] Am able to switch between jobs to set the active one
* [x] When I use the bin icon on the SQL Job Manager, all jobs end and are removed from the list
* [x] When I use the database icon on the SQL Job Manager, a new job is created and appears on the list
* [x] When I use the file icon with a plus on it, a new blank SQL file/tab is opened
* Running statements:
    * [x] When queries are run with an active job, they use the selected job (testing with `values (job_name)`)
    * [x] When queries are run with no active jobs, they use the default SQL API (which runs in a new job for every query)
    * [x] When navigating through the Schema Browser, queries are run through the active job in the SQL Job Manager
* Job Editor:
    * [x] When I use the pencil/edit button on an SQL Job, the properties window appears
    * [x] When I click save on the properties window, the old job is ended and a new one spins up with the new chosen settings from the properties view
    * [x] When I press the information icon on a job, the job log appears in a new tab
    * [x] When I click the bin icon, the job is ended and is removed from the list
    * [x] When I click the keyboard icon, the job ID is copied to my clipboard

File: src/connection/sqlJob.ts [hunk 1]
```diff
@@ -120,6 +122,34 @@ export class SQLJob {
...
+      id: `boop`,
+      type: `getversion`
+    };
+
+    const result = await this.send(JSON.stringify(verObj));
+
+    const version: VersionCheckResult = JSON.parse(result);
+
+    if (version.success !== true) {
+      throw new Error(version.error || `Failed to get version from packaned`);
```
Comment (davecharron) [inline src/connection/sqlJob.ts]: packaned?

File: src/views/jobManager/editJob/perfTab.ts [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+        selected: options["variable field compression"] === `false`,
+      },
+    ])
+    .addInput(
+      `query storage limit`,
+      `Query storage limit`,
+      `Specifies the query storage limit to be used when statements in a connection are executed. Valid values are -1 to 2147352578 megabytes. Note, this property is ignored when connecting to systems running IBM i V5R4 and earlier.
+    You must have *JOBCTL special authority to use query storage limit with Version 6 Release 1 of IBM i.`,
+      { default: options["query storage limit"] || `-1` }
+    );
```
Comment (davecharron) [inline src/views/jobManager/editJob/perfTab.ts:337]: Move up to line 311 to keep the query options together.

File: src/views/jobManager/editJob/perfTab.ts [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+        selected: options["query timeout mechanism"] === `qqrytimlmt`,
+      },
+    ])
+    // TODO: the followng two properties use a system default value, which is not supported by the current version of the driver:
+    // .addInput(`recieve buffer size`, `Recieve buffer size`, `	Specifies the buffer size used to receive data through the socket connection between the front-end driver and the IBM i system.
+    //   NOTE: This does not specify the actual receive buffer size. It is only used as a hint by the underlying socket code.`, {default: options["recieve buffer size"] || ``})
+    // .addInput(`send buffer size`, `Send buffer size`, `Specifies the buffer size used to send data through the socket connection between the front-end driver and the IBM i system.
+    //   NOTE: This does not specify the actual send buffer size. It is only used as a hint by the underlying socket code.`,
+    //   {default: options["send buffer size"] || ``})
+    .addSelect(`variable field compression`, `Variable field compression`, [
```
Comment (davecharron) [inline src/views/jobManager/editJob/perfTab.ts:317]: Missing the "use block update" property.
<html><body>


"use block update" | This property allows the toolbox JDBC driver to use a block      update mode when inserting or updating blocks of data into the database.        This is an optimized version of the batch insert or batch update. | no | "true"       "false" | "false"
-- | -- | -- | -- | --



</body>
</html>

File: src/views/jobManager/editJob/systemTab.ts [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+          selected: options["auto commit"] === `true`,
+        },
+        {
+          value: `false`,
+          description: `False`,
+          text: `No commit (false)`,
+          selected: options["auto commit"] === `false`,
+        },
+      ],
+      `Specifies whether auto-commit mode is the default connection mode for new connections. Calling AS400JDBCConnection.setAutoCommit(boolean) will override this property on a per-connection basis. Note that, in order to use transaction isolation levels other than *NONE when using auto-commit mode, the property "true autocommit" needs to be set to true.`
```
Comment (davecharron) [inline src/views/jobManager/editJob/systemTab.ts:58]:
```suggestion
      `Specifies whether auto-commit mode is the default connection mode for new connections. Note that, in order to use transaction isolation levels other than *NONE when using auto-commit mode, the property "true autocommit" needs to be set to true.`
```

File: src/views/jobManager/editJob/systemTab.ts [hunk 5]
```diff
@@ -0,0 +1,394 @@
...
+          selected: options["cursor sensitivity"] === `insensitive`,
+        },
+        {
+          value: `sensitive`,
+          description: `Sensitive`,
+          text: `sensitive`,
+          selected: options["cursor sensitivity"] === `sensitive`,
+        },
+      ],
+      formatDescription(autoCommitText)
```
Comment (davecharron) [inline src/views/jobManager/editJob/systemTab.ts:127]: Does this belong here?  Looks out of place.

File: src/views/jobManager/editJob/systemTab.ts [hunk 6]
```diff
@@ -0,0 +1,394 @@
...
+          selected: options["decfloat rounding mode"] === `up`,
+        },
+        {
+          value: `half down`,
+          description: `Half down`,
+          text: `half down`,
+          selected: options["decfloat rounding mode"] === `half down`,
+        },
+      ],
+      `Specifies the rounding mode to use when working with decfloat data type. Note, this property is ignored when connecting to systems running IBM i V5R4 and earlier.`
```
Comment (davecharron) [inline src/views/jobManager/editJob/systemTab.ts:179]:
```suggestion
      `Specifies the rounding mode to use when working with decfloat data type.`
```

File: src/views/jobManager/editJob/systemTab.ts [hunk 7]
```diff
@@ -0,0 +1,394 @@
...
+          selected: options["XA loosely coupled support"] === `0`,
+        },
+        {
+          value: `1`,
+          description: `1`,
+          text: `Share locks`,
+          selected: options["XA loosely coupled support"] === `1`,
+        },
+      ],
+      `Specifies whether lock sharing is allowed for loosely coupled transaction branches. Note, this property is ignored when connecting to systems running to IBM i V5R3 and earlier. This option can be set to 0 to indicate to not share locks or 1 to share locks.`
```
Comment (davecharron) [inline src/views/jobManager/editJob/systemTab.ts:384]:
```suggestion
      `Specifies whether lock sharing is allowed for loosely coupled transaction branches. This option can be set to 0 to indicate to not share locks or 1 to share locks.`
```

File: src/views/jobManager/editJob/perfTab.ts [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+        selected: options["prefetch"] === `false`,
+      },
+    ])
+    .addInput(
+      `qaqqinilib`,
+      `qaqqinilib`,
+      `	Specifies a QAQQINI library name. Used to specify the library that contains the qaqqini file to use. A qaqqini file contains all of the attributes that can potentially impact the performance of the DB2 for IBM i database engine. You must have *JOBCTL special authority to use qaqqinilib.`,
+      { default: options["qaqqinilib"] || `` }
+    )
+    .addSelect(`query optimize goal`, `Query optimize goal`, [
```
Comment (davecharron) [inline src/views/jobManager/editJob/perfTab.ts:277]: The description from the JDBC properties spec should be displayed.
> Specifies the goal the system should use with optimization of queries. This setting corresponds with the system's QAQQINI option called OPTIMIZATION_GOAL.

File: src/views/jobManager/jobLog.ts [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+              <vscode-data-grid-cell grid-column="${i++}">
+                ${row.MESSAGE_TYPE}
+              </vscode-data-grid-cell>
+              <vscode-data-grid-cell grid-column="${i++}">
+                ${row.MESSAGE_ID}
+              </vscode-data-grid-cell>
+              <vscode-data-grid-cell grid-column="${i++}">
+                ${escapeHTML(row.MESSAGE_TEXT)}
+              </vscode-data-grid-cell>
+            </vscode-data-grid-row>`
```
Comment (davecharron) [inline src/views/jobManager/jobLog.ts:59]: Should probably add a few more columns to the display, at a minimum

- SEVERITY
- MESSAGE_TIMESTAMP
- MESSAGE_SECOND_LEVEL_TEXT

File: src/views/jobManager/jobManagerView.ts [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+    context.subscriptions.push(
+      vscode.commands.registerCommand(`vscode-db2i.jobManager.refresh`, async () => {
+        this.refresh();
+      }),
+
+      vscode.commands.registerCommand(`vscode-db2i.jobManager.newJob`, async () => {
+
+        await window.withProgress({location: ProgressLocation.Window}, async (progress) => {
+          try {
+            progress.report({message: `Spinning up SQL job..`});
```
Comment (davecharron) [inline src/views/jobManager/jobManagerView.ts:25]:
```suggestion
            progress.report({message: `Spinning up SQL job...`});
```
Should we substitute `Starting` for `Spinning up`?

**Review by davecharron — CHANGES_REQUESTED** [review]
I've reviewed the code and have left comments and suggestions.

File: src/connection/serverComponent.ts [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+    if (this.outputChannel) {
+      this.outputChannel.appendLine(jsonString);
+    }
+  }
+
+  static getInitCommand(): string|undefined {
+    const path = this.getComponentPath();
+
+    if (path) {
+      return `/QOpenSys/QIBM/ProdData/JavaVM/jdk80/64bit/bin/java -jar ${path}`
```
[the remark this answers is not in the document]
Comment (ThePrez) [reply src/connection/serverComponent.ts:40]: Java 8 is guaranteed to be there and at this location, at least with all currently-released versions of IBM i.

File: package.json [hunk 12]
```diff
@@ -111,13 +111,27 @@
...
       ]
     },
     "viewsWelcome": [
       {
         "view": "queryHistory",
         "contents": "Query history will appear here."
+      },
+      {
+        "view": "jobManager",
+        "contents": "The SQL Job Manager allows you to manage multiple jobs and can each have their own connection settings, that you can run SQL statements in.\n\nUntil you have an SQL job running, all user run SQL statements will run in a new jobs.\n\nWant to run statements in a single, reusable job? \n[Start SQL job](command:vscode-db2i.jobManager.newJob)"
```
[the remark this answers is not in the document]
Comment (worksofliam) [reply package.json:134]: Beautiful. Much nicer than my informal approach.

File: src/connection/manager.ts [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+  async newJob(predefinedJob?: SQLJob) {
+    if (SQLJobManager.jobSupport) {
+      const instance = getInstance();
+      const config = instance.getConfig();
+
+      const newJob = predefinedJob || (new SQLJob({
+        libraries: [config.currentLibrary, ...config.libraryList], 
+        naming: `system`, 
+        "full open": false, 
+        "transaction isolation": "none"
```
[the remark this answers is not in the document]
Comment (worksofliam) [reply src/connection/manager.ts:28]: @ajshedivy Is this something you could set in a new commit to the branch?

File: src/views/jobManager/jobManagerView.ts (hunk 10 above)
Comment (worksofliam) [reply src/views/jobManager/jobManagerView.ts:25]: I like spinning up. Has a nice personal feeling to it.

File: src/views/jobManager/editJob/formatTab.ts [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+          selected: options["date separator"] === `,`,
+        },
+        {
+          value: `b`,
+          text: `space`,
+          description: `b`,
+          selected: options["date separator"] === `b`,
+        },
+      ],
+      `	Specifies the date separator used in date literals within SQL statements. This property has no effect unless the "date format" property is set to "julian", "mdy", "dmy" or "ymd".`
```
[the remark this answers is not in the document]
Comment (worksofliam) [reply src/views/jobManager/editJob/formatTab.ts:122]: Why the space at the beginning @davecharron?

File: src/views/jobManager/editJob/systemTab.ts (hunk 5 above)
Comment (worksofliam) [reply src/views/jobManager/editJob/systemTab.ts:127]: Yes, you're right. Cursor sensitivity doesn't belong in this instance.

@ajshedivy Let's remove this one.

**Comment by ThePrez** [comment]
> @ajshedivy I think we should delegate @ThePrez to determine if every JDBC Option needs testing. Seeing as we're using that database component, I would expect the database component to make sure they work.

We need to test that every JDBC option is specified/set correctly when it hits the JDBC driver. The database component is just passing through what VSC provides, so this should be verified at the VSC level. 

That being said, I have no idea how we'd go about testing that in a reliable, automated manner, apart from checking against the results of [getPropertyInfo](https://docs.oracle.com/javase/8/docs/api/java/sql/Driver.html#getPropertyInfo-java.lang.String-java.util.Properties-)

File: src/connection/manager.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (worksofliam) [reply src/connection/manager.ts:28]: @ThePrez I am going to set these properties. Will they have any impact I should know of?

File: src/views/jobManager/editJob/otherTab.ts [hunk 15]
```diff
@@ -0,0 +1,599 @@
...
+        },
+        {
+          value: `64`,
+          text: `start the database host server trace`,
+          description: `64`,
+          selected: options["server trace"] === `64`,
+        },
+      ],
+      `Specifies the level of tracing of the JDBC server job. When tracing is enabled, tracing starts when the client connects to the system and ends when the connection is disconnected. You must start tracing before connecting to the system, because the client enables system tracing only at connect time.`
+    )
```
[the remark this answers is not in the document]
Comment (worksofliam) [reply src/views/jobManager/editJob/otherTab.ts:433]: I don't think our UI is going to be capable of doing much for this right now and with that being said, we have another PR planned where we are going to handle tracing in the server component. It's not like this option will do very much nonetheless.

File: src/views/jobManager/jobLog.ts (hunk 9 above)
Comment (worksofliam) [reply src/views/jobManager/jobLog.ts:59]: I think we should do a seperate PR to improve the job log view. Right now, even with just 3 columns, there is a lot of data. I'd like to work on making it so we can toggle columns on and off. I created issue #26 to track that.

File: src/views/jobManager/editJob/otherTab.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (worksofliam) [reply src/views/jobManager/editJob/otherTab.ts:433]: Created #27 to track this.

File: src/views/jobManager/editJob/perfTab.ts (hunk 3 above)
Comment (worksofliam) [reply src/views/jobManager/editJob/perfTab.ts:324]: @davecharron I had a look into this but couldn't see it in the JDBC options docs online. Do you have docs for this?

https://www.ibm.com/docs/en/i/7.2?topic=ssw_ibm_i_72/rzahh/javadoc/com/ibm/as400/access/doc-files/JDBCProperties.htm

Comment (davecharron) [reply src/views/jobManager/editJob/perfTab.ts:324]: My bad, I was looking at an internal link.  I'm not finding the property in any externally available docs.

**Pull request merged**